src/json.c

```diff
@@ -159,6 +159,70 @@ int FieldSpec_CheckJsonType(FieldType fieldType, JSONType type, QueryError *stat
 //   return ret;
 // }
 
+#define BIT_CAST(type, value) ({ \
+  __typeof__(value) value_var = (value); \
+  static_assert(sizeof(type) == sizeof(value_var), "type and value must have the same size"); \
+  const void *ptr = &(value_var); \
+  *(type *)ptr; \
+})
+
+// Rounding the input to nearest even (float with 16 trailing zeros),
+// and returning the 16 significant bits of the float as uint16_t.
+// Calculation inspired by:
+// https://gitlab.com/libeigen/eigen/-/blob/d626762e3ff6cdcdf65325e6edf27c995029786d/Eigen/src/Core/arch/Default/BFloat16.h#L403
+static inline uint16_t floatToBF16bits(float input) {
+  uint32_t f32 = BIT_CAST(uint32_t, input);
+  uint32_t lsb = (f32 >> 16) & 1;
+  uint32_t round = lsb + 0x7FFF;
+  f32 += round;
+  return (f32 >> 16);
+}
+
+static int JSON_getBFloat16(RedisJSON json, uint16_t *val) {
+  double temp;
+  int ret = japi->getDouble(json, &temp);
+  if (REDISMODULE_OK == ret) {
+    *val = floatToBF16bits((float)temp);
+  }
+  return ret;
+}
+
+// via Fabian "ryg" Giesen.
+// https://gist.github.com/2156668
+static inline uint16_t floatToFP16bits(float input) {
+  const uint32_t sign_mask = 1u << 31;
+  const uint32_t f32infty = 255u << 23;
+  const uint32_t f16infty = 31u << 23;
+  const uint32_t round_mask = ~0xfffu;
+  const uint32_t magic_bits = 15u << 23;
+  const float magic = BIT_CAST(const float, magic_bits);
+
+  uint32_t f32 = BIT_CAST(uint32_t, input);
+  uint32_t sign = f32 & sign_mask;
+  f32 ^= sign;
+
+  uint32_t out;
+  if (f32 >= f32infty) {
+    out = f32 > f32infty ? 0x7e00u : 0x7c00u; // Inf or NaN
```

Comment:
should we have a similar inf/nan check for bfloat16?
can we cover this line?

Reply:
We don't validate the elements' values on the library and we assume it is valid (or at least we have undefined behavior if some of the elements are NaN or INF).
Since we don't have it anywhere else we can remove it from here as well